cratetorrent/src/tracker.rs

```diff
@@ -228,6 +301,152 @@ impl Tracker {
         let resp = serde_bencode::from_bytes(&resp)?;
         Ok(resp)
     }
+
+    async fn announce_udp(&self, params: Announce) -> Result<Response> {
+        let port = thread_rng().gen_range(1025..u16::MAX);
+        let mut sock = UdpSocket::bind(SocketAddr::from(([0, 0, 0, 0], port)))
+            .await
+            .unwrap();
+
+        // All of the potential addressese of a URL
+        let mut addrs = self.url.socket_addrs(|| None).unwrap();
+        // Shuffle the list
+        addrs.shuffle(&mut thread_rng());
+
+        //TODO: Make an error for not finding an actual IPV4 address
+        let addr = *addrs.iter().find(|a| a.is_ipv4()).unwrap();
+
+        let mut failure_reason = None;
+
+        let connection_id: i64 = Tracker::connect_udp(addr).await.unwrap();
+
+        const ACTION: i32 = 1;
+        let transaction_id: i32 = random();
+        let key: u32 = random();
+
+        let mut bytes_to_send = BytesMut::with_capacity(150);
+        bytes_to_send.put_i64(connection_id);
+        bytes_to_send.put_i32(ACTION);
+        bytes_to_send.put_i32(transaction_id);
+
+        debug_assert_eq!(params.info_hash.len(), 20);
+        debug_assert_eq!(params.peer_id.len(), 20);
+
+        bytes_to_send.put(&params.info_hash[..]);
+        bytes_to_send.put(&params.peer_id[..]);
+        bytes_to_send.put_i64(params.downloaded.try_into().unwrap());
+        bytes_to_send.put_i64(params.left.try_into().unwrap());
+        bytes_to_send.put_i64(params.uploaded.try_into().unwrap());
+
+        bytes_to_send.put_i32(match params.event {
+            Some(val) => match val {
+                crate::tracker::Event::Completed => 1,
+                crate::tracker::Event::Started => 2,
+                crate::tracker::Event::Stopped => 3,
+            },
+            None => 0,
+        });
+        match params.ip {
+            Some(ip) => {
+                match ip {
+                    IpAddr::V4(ip) => {
+                        bytes_to_send.put(&(ip.octets()[..]));
+                    }
+
+                    //The IP address field must be 32 bits wide, so if the IP given is v6, the field must be set to 0
+                    IpAddr::V6(_) => {
+                        bytes_to_send.put_i32(0);
+                    }
+                }
+            }
+            None => {
+                bytes_to_send.put_i32(0);
+            }
+        };
+
+        bytes_to_send.put_u32(key);
+        bytes_to_send.put_i32(match params.peer_count {
+            Some(num) => num.try_into().unwrap(),
+            None => -1,
+        });
+        bytes_to_send.put_u16(params.port);
+
+        let bytes_to_send = &bytes_to_send;
+
+        const MAX_NUM_PEERS: usize = 2048;
+
+        //Supporting around a few hundred peers, just for the test
+
+        let mut response_buf: [u8; MAX_NUM_PEERS] = [0; MAX_NUM_PEERS];
```

Comment:
We should support an arbitrary number of peers.

Reply:
I could probably just turn it into a vector if you'd like